### PR TITLE
ore: Add new `ore::panic::catch_unwind_str` function

### PR DESCRIPTION
**Pull request by ParkMyCar**
Today `ore::panic::catch_unwind` returns a `Box<dyn Any>` which isn't useful to callers since the `Debug` format when printing is just `Any`. This PR adds a new methond `catch_unwind_str` that tries downcasting to a `&'static str` or `String` which is the type of the returning object should be, and is the panic message.

### Motivation

Happened upon this while talking to @frankmcsherry 

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Review by benesch — COMMENTED** [review]
I know this is only called in one place right now but forcing the downcast to `<String>` seems too specific for a low-level utility crate like ore. Definitely a good idea for that one callsite though. Can you either put this downcasting at the call site, or introduce a new `catch_unwind_string` method in ore?

**Comment by ParkMyCar** [comment]
> I know this is only called in one place right now but forcing the downcast to `<String>` seems too specific for a low-level utility crate like ore. Definitely a good idea for that one callsite though. Can you either put this downcasting at the call site, or introduce a new `catch_unwind_string` method in ore?

Makes sense, added a new `catch_unwind_str` method!

**Review by benesch — APPROVED** [review]
Looks great, thank you!

File: src/ore/src/panic.rs [hunk 1]
```diff
@@ -196,3 +197,23 @@ where
         res
     })
 }
+
+/// Like [`crate::panic::catch_unwind`], but downcasts the returned `Box<dyn Any>` error to a
+/// string which is almost always is.
```
Comment (antiguru) [inline src/ore/src/panic.rs:202]: which _it_ almost

:)

**Pull request merged**